.. _tools:

MoMI-G tools
===================

The MoMI-G package includes a set of scripts (MoMI-G tools) that converts a VCF file into the variation graph format. 

.. image:: tools.*

If you want to run MoMI-G with your own dataset, use our custom scripts ``scripts/vcf2xg.sh`` to generate pcf and xg dataset. It requires VG, ruby, bash, and samtools.

.. csv-table::
   :header: Software, Dataset, Supported SV type
   :widths: 20, 20, 40

   Sniffles, for PacBio/Nanopore, INV/DEL/DUP/TRA/(INS)
   10X LongRanger, for 10X, To be supported
   SURVIVOR, for merging SV calls, To be supported


.. code-block:: console

  $ bash vcf2xg.sh test.vcf test_output /bin/vg hg[19|38]

After that, these files are required to be mounted on ``static/`` folder. Also, you should modify ``config.yaml``.

* static/
    * config.yaml: a configuration file
    * \*.xg: an index of a variation graph, generated by [vg](https://github.com/vgteam/vg)
    * \*.pcf: a pair of cooridnate format file: required to display variants on Feature Table or Circos Plot. 
    * \*.gam(optional): read alignments on the graph
    * \*.gam.index(optional): index of gam

Then, run the MoMI-G frontend.

.. code-block:: console

  $ docker run --init -p 8081:8081 -v `pwd`/static:/vg/static momigteam/momig-backend
  $ yarn start


